File: onyx-database/src/com/onyx/fetch/impl/RelationshipScanner.java
package com.onyx.fetch.impl;

import com.onyx.descriptor.EntityDescriptor;
import com.onyx.descriptor.RelationshipDescriptor;
import com.onyx.exception.EntityException;
import com.onyx.fetch.PartitionReference;
import com.onyx.fetch.ScannerFactory;
import com.onyx.fetch.TableScanner;
import com.onyx.persistence.context.SchemaContext;
import com.onyx.persistence.manager.PersistenceManager;
import com.onyx.persistence.query.Query;
import com.onyx.persistence.query.QueryCriteria;
import com.onyx.record.RecordController;
import com.onyx.relationship.RelationshipController;
import com.onyx.relationship.RelationshipReference;
import com.onyx.structure.MapBuilder;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Created by timothy.osborn on 1/3/15.
 */
public class RelationshipScanner extends AbstractTableScanner implements TableScanner {

    protected RelationshipDescriptor relationshipDescriptor;

    /**
     * Constructor
     *
     * @param criteria
     * @param classToScan
     * @param descriptor
     */
    public RelationshipScanner(QueryCriteria criteria, Class classToScan, EntityDescriptor descriptor, MapBuilder temporaryDataFile, Query query, SchemaContext context, PersistenceManager persistenceManager) throws EntityException
    {
        super(criteria, classToScan, descriptor, temporaryDataFile, query, context, persistenceManager);

    }

    /**
     * Full Table get all relationships
     *
     */
    @Override
    public Map<Long, Long> scan() throws EntityException
    {
        return scan(records);
    }

    /**
     * Full Scan with existing values
     *
     * @param existingValues
     * @return
     * @throws EntityException
     */
    @Override
    public Map scan(Map existingValues) throws EntityException
    {
        // Retain the original attribute
        final String originalAttribute = criteria.getAttribute();

        // Get the attribute name.  If it has multiple tokens, that means it is another relationship.
        // If that is the case, we gotta find that one
        final String[] segments = originalAttribute.split("\\.");

        // Map <ChildIndex, ParentIndex> // Inverted list so we can use it to scan using an normal full table scanner or index scanner
        final Map relationshipIndexes = getRelationshipIndexes(segments[0], existingValues);
        final Map returnValue = new HashMap();

        // We are going to set the attribute name so we can continue going down the chain.  We are going to remove the
        // processed token through
        criteria.setAttribute(criteria.getAttribute().replaceFirst(segments[0] + "\\.", ""));

        // Get the next scanner because we are not at the end of the line.  Otherwise, we would not have gotten to this place
        final TableScanner tableScanner = ScannerFactory.getInstance(getContext()).getScannerForQueryCriteria(criteria, relationshipDescriptor.getInverseClass(), temporaryDataFile, query, persistenceManager);

        // Sweet, lets get the scanner.  Note, this very well can be recursive, but sooner or later it will get to the
        // other scanners
        final Map childIndexes = tableScanner.scan(relationshipIndexes);

        // Swap parent / child after getting results.  This is because we can use the child when hydrating stuff
        for (Object childIndex : childIndexes.keySet())
        {
            // Gets the parent
            returnValue.put(relationshipIndexes.get(childIndex), childIndex);
        }

        criteria.setAttribute(originalAttribute);

        return returnValue;
    }

    /**
     * Get Relationship Indexes
     *
     * @param attribute
     * @param existingValues
     * @return
     * @throws EntityException
     */
    protected Map getRelationshipIndexes(String attribute, Map existingValues) throws EntityException
    {
        final Map allResults = new HashMap();

        final Iterator iterator = existingValues.keySet().iterator();

        relationshipDescriptor = descriptor.getRelationships().get(attribute);
        final RelationshipController relationshipController = getContext().getRelationshipController(relationshipDescriptor);
        final RecordController inverseRecordController = getDefaultInverseRecordController();

        List<RelationshipReference> relationshipIdentifiers = null;
        Object keyValue = null;

        while(iterator.hasNext())
        {
            if(query.isTerminated())
                return allResults;

            keyValue = iterator.next();

            if(keyValue instanceof PartitionReference)
            {
                relationshipIdentifiers = relationshipController.getRelationshipIdentifiersWithReferenceId((PartitionReference)keyValue);
            }
            else
            {
                relationshipIdentifiers = relationshipController.getRelationshipIdentifiersWithReferenceId((long)keyValue);
            }
            for(RelationshipReference id : relationshipIdentifiers)
            {

                if(id.partitionId == 0)
                {
                    allResults.put(inverseRecordController.getReferenceId(id.identifier), keyValue);
                }
                else
                {
                    RecordController recordController = getRecordControllerForPartition(id.partitionId);
                    PartitionReference reference = new PartitionReference(id.partitionId, recordController.getReferenceId(id.identifier));
                    allResults.put(reference, keyValue);
                }

            }
        }

        return allResults;
    }

    /**
     * Grabs the inverse record controller
     *
     * @return
     * @throws EntityException
     */
    protected RecordController getDefaultInverseRecordController() throws EntityException
    {
        final EntityDescriptor inverseDescriptor = getContext().getBaseDescriptorForEntity(relationshipDescriptor.getInverseClass());
        return getContext().getRecordController(inverseDescriptor);
    }
}

File: onyx-database/src/main/kotlin/com/onyx/interactors/scanner/impl/RelationshipScanner.kt
package com.onyx.interactors.scanner.impl

import com.onyx.descriptor.EntityDescriptor
import com.onyx.exception.OnyxException
import com.onyx.exception.InvalidQueryException
import com.onyx.extension.*
import com.onyx.extension.common.instance
import com.onyx.interactors.record.data.Reference
import com.onyx.interactors.scanner.ScannerFactory
import com.onyx.interactors.scanner.TableScanner
import com.onyx.persistence.IManagedEntity
import com.onyx.persistence.context.SchemaContext
import com.onyx.persistence.manager.PersistenceManager
import com.onyx.persistence.query.Query
import com.onyx.persistence.query.QueryCriteria
import com.onyx.persistence.query.QueryPartitionMode
import com.onyx.persistence.context.Contexts

/**
 * Created by timothy.osborn on 1/3/15.
 *
 * Scan relationships for matching criteria
 */
class RelationshipScanner @Throws(OnyxException::class) constructor(criteria: QueryCriteria, classToScan: Class<*>, descriptor: EntityDescriptor, query: Query, context: SchemaContext, persistenceManager: PersistenceManager) : AbstractTableScanner(criteria, classToScan, descriptor, query, context, persistenceManager), TableScanner {

    /**
     * Full Table get all relationships
     *
     */
    @Throws(OnyxException::class)
    override fun scan(): MutableSet<Reference> {

        val context = Contexts.get(contextId)!!

        // We do not support querying relationships by all partitions.  That would be horribly inefficient
        if (this.query.partition === QueryPartitionMode.ALL) throw InvalidQueryException()

        var partitionId = 0L

        if (this.descriptor.hasPartition) {
            val temporaryManagedEntity: IManagedEntity = descriptor.entityClass.instance(this.contextId)
            temporaryManagedEntity[context, descriptor, descriptor.partition!!.name] = query.partition.toString()
            partitionId = temporaryManagedEntity.partitionId(context, descriptor)
        }


        return scan(records.references.map { Reference(partitionId, it.position) }.toHashSet())
    }

    /**
     * Full Scan with existing values
     *
     * @param existingValues Existing values to check criteria
     * @return filtered map of results matching additional criteria
     * @throws OnyxException Cannot scan relationship values
     */
    @Throws(OnyxException::class)
    override fun scan(existingValues: Set<Reference>): MutableSet<Reference> {
        val context = Contexts.get(contextId)!!
        // Get the next scanner because we are not at the end of the line.  Otherwise, we would not have gotten to this place
        val tableScanner = ScannerFactory.getFullTableScanner(context, criteria, descriptor.entityClass, query, persistenceManager)
        if(isLast) {
            tableScanner.isLast = true
            this.collector = tableScanner.collector
        }

        // Sweet, lets get the scanner.  Note, this very well can be recursive, but sooner or later it will get to the
        // other scanners
        return tableScanner.scan(existingValues)
    }

}
